feat: add scrollbar widget

kdheepak opened the pull request:
I've added a simple scrollbar widget.

```
 <--▮------->
 ^  ^   ^   ^
 │  │   │   └ end arrow
 │  │   └──── track symbol
 │  └──────── thumb symbol
 └─────────── begin arrow
```

<img width="748" alt="image" src="https://github.com/tui-rs-revival/ratatui/assets/1813121/ca361775-1163-4e74-8e39-f656c0325695">

It can be used to add a scrollbar to any `Rect`. 
It can also be customized with different styles and symbols for, the bar, the "scrollwheel" and arrows.

This addresses https://github.com/tui-rs-revival/ratatui/issues/173

A larger discussion is to be had about scrollable widgets in general: https://github.com/tui-rs-revival/ratatui/issues/174

I'm open to feedback on improving this. I've needed a scrollbar for a couple of projects now and figured I'd clean up the code and share it as a `Widget` here. 

_Docstrings were generated using ChatGPT_.

joshka commented on src/widgets/scrollbar.rs:
```diff
@@ -0,0 +1,208 @@
+    }
+
+    /// Sets the symbol that represents the scroller.
+    pub fn symbol(mut self, symbol: &'a str) -> Self {
```
I wonder if we should just use https://crates.io/crates/derive_setters for new widgets.

kdheepak replied:
I'm assuming this can be done as a separate PR for all widgets, so I'll resolve this conversation for now.

joshka replied:
yep

joshka commented on src/widgets/scrollbar.rs:
```diff
@@ -0,0 +1,208 @@
+            buf.set_string(right, y, self.bar_symbol.clone(), self.bar_style);
+        }
+
+        // Adjust the scroll position to fit within the range.
+        let scroll = if self.scroll >= 1.0 {
+            // Conversions rounds down, so make this slightly less than 1.0
+            0.9999
+        } else {
+            self.scroll
+        };
+        // Calculate the position of the scroller based on the scroll position and the height of the
+        // scrollbar.
+        let pos = height as f64 * scroll;
```
can you use `clamp()` here? e.g. something like:
 `let pos = (height * self.scroll).clamp(0, height) as u16;`?

kdheepak replied:
This is not required with the new implementation using `position` and `num_elements`.

joshka commented on src/widgets/scrollbar.rs:
```diff
@@ -0,0 +1,208 @@
+    ///
+    /// The scrollbar is drawn with a fixed width on the right edge of the given area.
+    fn render(self, area: Rect, buf: &mut Buffer) {
+        // If the area is too small, we don't draw the scrollbar.
+        if area.height <= 2 {
```
nit: use a constant to avoid writing a comment (e.g. `MIN_HEIGHT`)

kdheepak replied:
Do you mean writing it like this?

```
        const MIN_HEIGHT: u16 = 0;

        if area.height == MIN_HEIGHT {
            return;
        }
```

joshka replied:
More try to explain why the min height is 2 in a constant - here's it's because we're subtracting the top and the bottom row and not rendering them. Something like
```rust
// subtract space for top and bottom
let Some(bar_height) = area.height.checked_sub(2) or else { return; };
```

I think this has disappeared in the current iteration though (which checks for 0 - which is obvious and fine).

joshka commented on src/widgets/scrollbar.rs:
```diff
@@ -0,0 +1,208 @@
+        // Calculate the rightmost column for the scrollbar.
+        let right = area.right().saturating_sub(1);
+
+        // If the rightmost column is out of bounds, we don't draw the scrollbar.
+        if right < area.left() {
+            return;
+        };
```
nit: handle this in the same if statement above (as `MIN_WIDTH`)

joshka commented on src/widgets/scrollbar.rs:
```diff
@@ -0,0 +1,208 @@
+
+        // Draw the bar of the scrollbar.
+        for y in top..(top + height) {
+            buf.set_string(right, y, self.bar_symbol.clone(), self.bar_style);
```
what's the clone() issue that github flags here?

kdheepak replied:
I previously had `self.bar_symbol` as a `String` in my personal project but changed it to `&str` before submitting the PR, but I didn't remove the `.clone()`. I think with `&str` the `.clone()` is not required. With the `-D` flag, we get this error:

```
error: using `clone` on a double-reference; this will copy the reference of type `&str` instead of cloning the inner type
Error:   --> src/widgets/scrollbar.rs:91:38
   |
91 |             buf.set_string(right, y, self.bar_symbol.clone(), self.bar_style);
   |                                      ^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#clone_double_ref
   = note: `#[deny(clippy::clone_double_ref)]` on by default
help: try dereferencing it
   |
91 |             buf.set_string(right, y, &(*self.bar_symbol).clone(), self.bar_style);
   |                                      ~~~~~~~~~~~~~~~~~~~~~~~~~~~
help: or try being explicit if you are sure, that you want to clone a reference
   |
91 |             buf.set_string(right, y, <&str>::clone(self.bar_symbol), self.bar_style);
   |                                      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

joshka commented:
Thanks for working on this btw - looking forward to seeing it be part of ratatui.

kdheepak commented:
Thanks for the detailed feedback!!! 

Let me address some of the comments and get back to you when it is ready for another round for review. 

I'm also planning to update this scrollbar widget so that it works horizontally as well.

kdheepak commented on src/widgets/scrollbar.rs:
```diff
@@ -0,0 +1,208 @@
+/// This widget can be used to display a scrollbar in a terminal user interface.
+#[derive(Default, Debug, Clone)]
+pub struct Scrollbar<'a> {
+    scroll: f64,
```
I've updated the code to make it show the size relative to the "number of lines" that the scroll bar can scroll over.
Previously, `scroll` was a ratio of `position / num_elements`. Now, the widget requires a user to pass in both `position` and `num_elements` explicitly.

joshka replied:
I think it could be useful to have multiple approaches to this (like the gauge does - it allows ratio, percent, etc.)

I'm not sure that I understand what num_elements represents just from the name:
1. The number of lines of the bar part of the scrollbar 
2. The total number of elements that are being scrolled (e.g. 1000 list items)
3. The total number of lines in the elements that are being scrolled (e.g. 1000 list items * 3 lines each)
4. The number of elements that are currently being shown (10 list items)
5. The number of lines that are currently being shown (30 lines)

kdheepak commented on src/widgets/scrollbar.rs:
```diff
@@ -0,0 +1,208 @@
+        let (top, height) = { (area.top() + 1, area.height.saturating_sub(2)) };
+
+        // Draw the bar of the scrollbar.
+        for y in top..(top + height) {
```
I've modified it to render to the entire width or height of the rectangle now.

kdheepak commented on src/widgets/scrollbar.rs:
```diff
@@ -0,0 +1,208 @@
+            })
+            .unwrap();
+
+        terminal.backend().assert_buffer(&expected);
```
Thanks for the suggestion on how to do tests. I've updated all the tests here to use your suggestion.

joshka commented on src/widgets/scrollbar.rs:
```diff
@@ -0,0 +1,208 @@
+        Self {
+            scroll,
+            style: Style::default(),
+            symbol: FULL,
+            bar_style: Style::default(),
+            bar_symbol: DOUBLE_VERTICAL,
```
The context on this one is consistency with the rest of the crate (e.g. the bar_sets idea).

kdheepak replied:
I've taken an initial stab at an interface with a `Set` in https://github.com/tui-rs-revival/ratatui/pull/228/commits/ab154507fc0d0dbc9ebdc63d10fa2fb08608136f. Let me know what you think.

kdheepak replied:
I believe this has been addressed now. I followed the same pattern as in bar::Set.

joshka commented on src/widgets/scrollbar.rs:
```diff
@@ -4,11 +4,54 @@ use crate::{
     widgets::Widget,
 };
 
+/// Scrollbar Set
+/// <--▮------->
+/// ^  ^   ^   ^
+/// │  │   │   └ end
+/// │  │   └──── track
+/// │  └──────── thumb
+/// └─────────── begin
```
This is a neat way to show this.

joshka commented on src/widgets/scrollbar.rs:
```diff
@@ -67,35 +120,79 @@ impl<'a> Scrollbar<'a> {
+    pub fn arrows(mut self, arrows: bool) -> Self {
+        self.arrows = arrows;
         self
     }
```
Do you think we could omit this and make start and end `Optional<&str>` where there's nothing rendered if they're set to none?

kdheepak replied:
Do we want to support arrows just at the beginning or just at the end, if one is set to `Some` and the other is set to `None`? The rendering logic gets more complicated in that case.

joshka replied:
Perhaps we could have an `arrows(start: &str, end: &str)` method?

kdheepak replied:
I've added a version in https://github.com/tui-rs-revival/ratatui/pull/228/commits/aec39611820f4cbbea501fbcbad498444d2b3790 and https://github.com/tui-rs-revival/ratatui/pull/228/commits/e6a68954245ccf2eee1c9f7354dc195b77e0a49f that uses `Option<&str>` for the begin and end symbols.

kdheepak replied:
I believe this comment has been addressed now.

joshka commented on src/widgets/scrollbar.rs:
```diff
@@ -0,0 +1,466 @@
+    }
+
+    #[test]
+    fn render_scrollbar_two_size_horizontal() {
```
neat!

joshka reviewed (commented):
Nice overall

I'd still prefer to have the position part out of the scroll bar - based on the stuff in the other PR / design Issue, but this is really good. I'll take a full review sometime soon.

kdheepak commented on src/widgets/scrollbar.rs:
```diff
@@ -0,0 +1,208 @@
+    }
+
+    /// Sets the scroll position of the scrollbar.
+    pub fn scroll(mut self, scroll: f64) -> Self {
```
I've not really worked with stateful widgets all that much, so I'm still trying to wrap my head around it. I'll have to review the code in your PR to get a sense of what is required here.

kdheepak replied:
I took a crack at it in https://github.com/tui-rs-revival/ratatui/pull/228/commits/739693c1ca879f09d01bd3885d02e2821dc92e88 and https://github.com/tui-rs-revival/ratatui/pull/228/commits/7f410a6dd890435d0581ebd022859a623e24fb73.

Is a StatefulWidget mainly useful just to "separate concerns"? Are there other practical reasons one would want to choose a StatefulWidget over a stateless widget?

joshka replied:
I think mostly separating the state means that the widget can modify the state during rendering. That’s necessary if you want to track what’s visible.

imagine we just want to render a scroll bar that responds to arrow keys. We can manually track the scroll offset etc, and handle edge cases or just call scroll_state.up()/down() and just pass that to the scroll bar. All the logic about how many items there are etc is part of the state. The scroll bar then simply handles the rendering.

kdheepak replied:
I've updated to use a State. In this case, we only need one direction because currently we are not rendering a 2D scroll bar.

kdheepak commented on src/widgets/scrollbar.rs:
```diff
@@ -0,0 +1,606 @@
+            thumb_style: Style::default(),
+            track_symbol: DOUBLE_VERTICAL.track,
+            track_style: Style::default(),
+            begin_symbol: None, // Some(DOUBLE_VERTICAL.begin),
+            begin_style: Style::default(),
+            end_symbol: None, // Some(DOUBLE_VERTICAL.end),
+            end_style: Style::default(),
```
Should arrows be enabled by default?

kdheepak replied:
I'm leaning toward enabling arrows by default. 

One reason is when using a scrollbar with the same size of the rect, the arrows take up the corners:

<img width="733" alt="image" src="https://github.com/tui-rs-revival/ratatui/assets/1813121/d051c1b8-c6a8-4f6c-b9ea-a05a784db763">

when is a better out of the box experience than this:

<img width="732" alt="image" src="https://github.com/tui-rs-revival/ratatui/assets/1813121/f020bc25-8753-4913-9fe1-f11166fd30ec">

Without arrows, most users will want to write this:

```
        Scrollbar::default()
            .orientation(ScrollbarOrientation::Vertical)
            .arrows(false),
        chunks[0].inner(&Margin {
            vertical: 1,
            horizontal: 0,
        })
```

i.e. decrease the size of the scrollbar by 1 so that it looks nicer:

<img width="726" alt="image" src="https://github.com/tui-rs-revival/ratatui/assets/1813121/ac130c09-1f80-40d3-be6d-812853762fc0">

That is a less nice first experience for a user.

joshka replied:
+1 to arrows by default.

I'd be inclined to say that for the scrollbar to render as part of a paragraph that has borders, it should be inside the borders not on the borders (but also that this should be configurable - I like the idea of being able to use the border as a scroll bar like what you have here).

kdheepak replied:
The scrollbar widget here doesn’t know anything about other widgets. 

Are you imaging that another widget like a  `Paragraph` would store an instance to a scrollbar and manipulate where it is being rendered?

kdheepak replied:
Btw, I like the scrollbar in the border because it reuses space that isn’t useful (inside it may overwrap on text or eat into the width or height of a text area), and because it can be turned off and on without any reflow on content in the widget.

joshka replied:
yup - I'm thinking about how existing widgets will use the scrollbar.
The would create one in their render() method.

kdheepak replied:
That makes sense! Happy to add any other functions you think will improve the interface for other widgets or users.

kdheepak commented:
I think this is ready for a final review now!

sayanarijit commented:
Quick glance looks good.

kdheepak commented on src/widgets/scrollbar.rs:
```diff
@@ -0,0 +1,698 @@
+    }
+
+    #[test]
+    fn render_scrollbar_two_size_double_vertical() {
```
I'm looking at all the test function names in this PR and they are not good at conveying the intent of the test.

I came across this article a few days ago:

https://bitfieldconsulting.com/golang/test-names

This recommends having test function names be full sentences. For example, in this case the test function name would be:

```rust
fn when_height_is_twice_content_length_then_thumb_size_is_length_two() {
```

Thoughts?

kdheepak replied:
I've refactored all the test names so hopefully the tests will be more clearer now.

joshka replied:
I've used a number of different conventions in various programming languages over the years. In the rust code I've read so far, there doesn't seem to be a wide consensus on naming (or in many cases doing the normal ActArrangeAssert style or GivenWhenThen style tests). I'd go with the idea that the name of the test should give a developer that sees it enough information to understand the intent of what is being tested. Test names help as a way of memorializing the design decisions, or by characterizing the behavior of edge cases.

If you take that approach, then a name like `thumb_size_is_proportional_to_content_length()` might be the descriptive part that captures the requirement / intent. Saying the parts about "twice" and "length two" are implementation details of that broader requirement. You can get that from reading the test, but you can't easily see why you care about height being twice content length without a test name, assertion message, or doc comment (or all three).

kdheepak replied:
I took another stab at better test names in https://github.com/tui-rs-revival/ratatui/pull/228/commits/0dc6770505dbdaff3f2b81d586207cbb970e10d2 and https://github.com/tui-rs-revival/ratatui/pull/228/commits/c436716621e969b7e299b00b0f23797e8277e60e.

kdheepak commented on src/widgets/scrollbar.rs:
```diff
@@ -0,0 +1,698 @@
+            ScrollbarOrientation::Vertical => DOUBLE_VERTICAL,
+            ScrollbarOrientation::Horizontal => DOUBLE_HORIZONTAL,
+        })
+    }
```
> `ScrollBar::default().orientation(Orientation::Horizontal) != ScrollBar::new(Orientation::Horizontal)`

Why do you say that this is the case? Aren't they identical?

I would say that majority of the users are going to want to just choose between a Vertical or Horizontal scrollbar, so we want to make that "path" easy to use. Which is why I think setting orientation should change the symbols used.

joshka replied:
Yep makes sense - I just misread the way the interaction between new and default worked.
I'm used to seeing new actually returning Self { ... } and assumed wrongly that it would be setting the orientation directly on the new struct, but it's actually creating it in default() and then setting the orientation with the builder method. This is fine I think.

kdheepak commented on src/widgets/scrollbar.rs:
```diff
@@ -0,0 +1,698 @@
+    pub begin_style: Style,
+    pub end_symbol: Option<&'a str>,
+    pub end_style: Style,
+    pub mirror: bool,
```
You are right, they don't. 

Resolved in https://github.com/tui-rs-revival/ratatui/pull/228/commits/c26cd1fe14dac16858f99a4af76ed6b10b992362.

kdheepak commented on src/widgets/scrollbar.rs:
```diff
@@ -0,0 +1,698 @@
+    pub fn symbol(mut self, symbol: Set) -> Self {
+        self.track_symbol = symbol.track;
+        self.thumb_symbol = symbol.thumb;
+        if self.begin_symbol.is_some() {
+            self.begin_symbol = Some(symbol.begin);
+        }
+        if self.end_symbol.is_some() {
+            self.end_symbol = Some(symbol.end);
+        }
```
I've added a few lines to the docstring. 

```
    /// Only sets begin_symbol and end_symbol if they are already contain a value.
    /// If begin_symbol and/or end_symbol were set to `None` explicitly, this function will respect
    /// that choice.
```

Hopefully that helps clarify it?

kdheepak commented on src/widgets/scrollbar.rs:
```diff
@@ -0,0 +1,698 @@
+            self = self.end_symbol(None);
+        }
+        self
+    }
```
I decided to remove the `arrows` method. 

So now, this will draw arrows:

```rust
            Scrollbar::default()
```

And this will not draw arrows:

```rust
            Scrollbar::default()
                .begin_symbol(None)
                .end_symbol(None)
```

kdheepak commented on src/widgets/scrollbar.rs:
```diff
@@ -0,0 +1,698 @@
+    pub fn mirror(mut self, mirror: bool) -> Self {
+        self.mirror = mirror;
+        self
+    }
```
I've removed the `mirror` method in favor of the orientation enum capturing the location of the scrollbar.

kdheepak commented on src/widgets/scrollbar.rs:
```diff
@@ -0,0 +1,698 @@
+impl<'a> StatefulWidget for Scrollbar<'a> {
+    type State = ScrollbarState;
+
+    fn render(self, area: Rect, buf: &mut Buffer, state: &mut Self::State) {
```
Good idea. I've broken it down in this commit:

https://github.com/tui-rs-revival/ratatui/pull/228/commits/cd579d703c8eb38928b9ad7a0cd6ba316144f98e.

kdheepak commented on src/widgets/scrollbar.rs:
```diff
@@ -0,0 +1,698 @@
+        // info!("{:?}", area);
+
+        // Define main size, cross position, and main position based on scrollbar orientation
+        let (main_size, main_position, cross_position, other_cross_position) =
```
I've added more documentation to the code to explain this.

```
        // `main` is the axis along which the scrollbar is being rendered.
        // `cross` is the other axis.
        //
        // For ScrollbarOrientation::VerticalRight
        //
        //        ┌───────────────────────── other_cross_position
        //        │               ┌───────── cross_position
        //        v               v
        //        ┌───────────────┐  ┌────── main_position
        //        │               ║<─┘ \
        //        │               █    }
        //  main  │               █    } ─── main_size
        //        │               ║    }
        //        │               ║    /
        //        └───────────────┘
        //               cross
        //
        // For ScrollbarOrientation::HorizontalBottom
        //
        //              main
        //        ┌───────────────┐<──────── other_cross_position
        //        │               │
        //  cross │               │
        //        │               │
        //        └═══███═════════┘<──────── cross_position
        //         ^
        //         │
        //         └──────────────────────── main_position
        //         \~~~~~~~~~~~~~/
        //                 └──────────────── main_size

```

kdheepak commented on src/widgets/scrollbar.rs:
```diff
@@ -0,0 +1,698 @@
+
+        // Define main size, cross position, and main position based on scrollbar orientation
+        let (main_size, main_position, cross_position, other_cross_position) =
+            if is_vertical && self.mirror {
```
I moved it to a separate method and added a comment. Along with the new enum that captures the orientation, hopefully it'll be clearer.

kdheepak commented on src/widgets/scrollbar.rs:
```diff
@@ -0,0 +1,698 @@
+            || (!(self.mirror) && cross_position <= other_cross_position)
+        {
+            return;
+        }
```
I previously chose to not render the scrollbar vertically if the width == 1 but I think that is not needed. I removed it now and it will render a vertical scrollbar even if the width of the rect is 1.

kdheepak commented on src/widgets/scrollbar.rs:
```diff
@@ -0,0 +1,698 @@
+            return;
+        }
+
+        // Calculate the scroll ratio based on the current position over the length of
+        // scrollable content.
+        let scroll_position_ratio = (state.position as f64 / state.content_length as f64).min(1.0);
+
+        // Calculate the scrollbar size based on the length of scrollable content, using
+        // a proportion of the total area main size.
+        let scrollbar_size = (((main_size as f64 / state.content_length as f64) * main_size as f64)
+            .round() as u16)
+            .max(1);
```
I've refactored this into a new method.

kdheepak commented on src/widgets/scrollbar.rs:
```diff
@@ -0,0 +1,698 @@
+    }
+
+    #[test]
+    fn render_scrollbar_position_horizontal_arrows_mirror() {
```
I made the ScrollbarOrientation like so:

```
/// Scrollbar Orientation
#[derive(Default, Debug, Clone)]
pub enum ScrollbarOrientation {
    #[default]
    VerticalRight,
    VerticalLeft,
    HorizontalBottom,
    HorizontalTop,
}
```

This removes the need for mirror as method and option.

kdheepak commented on src/widgets/scrollbar.rs:
```diff
@@ -0,0 +1,698 @@
+        // doesn't have room to draw, or if length of the contents to be scrolled is zero.
+        if main_size == 0 || state.content_length == 0 {
+            return;
+        }
```
I'm not sure I follow what you mean by this comment unfortunately. Can you clarify?

kdheepak commented on src/widgets/scrollbar.rs:
```diff
@@ -0,0 +1,698 @@
+/// │  │   └──── track
+/// │  └──────── thumb
+/// └─────────── begin
+#[derive(Debug, Clone)]
```
The current render function allows just a begin arrow or just a end arrow, and that is done by checking whether or not there's an `Some` or `None` for those symbols. But the `Set` stores all the values, so we'd have to store additional variables to check for those conditions. The current implementation seems cleaner to me.

kdheepak commented:
@joshka I've addressed a number of your comments and asked for some clarification on a new questions.

joshka reviewed (changes requested):
Picking back up the conversation about the cross. The code doesn't make it clear how the cross (perhaps you mean corner?) is relevant to the the algorithm for rendering the scroll_bar.

If I'm thinking through this from basics, 200 lines to render a scrollbar seems excessive.
It's possible that there are some assumptions or non obvious edge cases that you have factored in here, but they are difficult to see. I'm not sure it would be easy to fix a problem in this code, or to work out how to add a new feature.

I think perhaps this works too much in absolute coordinates and worries about the vertical / horizontal distinction throughout rather than using a much simpler relative approach.

An algorithm that treats the rendering as rendering a line from 0 to bar length would look like:
```rust
// this is psuedo rust, please forgive any errors
let length = length_of_bar_without_arrows(); // e.g. either area.width or area.height -1 or -2 
let (start_track, end_track) = relative_positions_of_track_without_arrows(); // e.g. (0, length - 1) or (1, length -2);
let (start_thumb, end_thumb) = relative_positions_of_thumb(); // e.g. (3, 6)

// this can be a method... absolute_positions()
let (left, top, right, bottom) =
  (area.x, area.y, area.x + area.width - 1, area.y + area.height - 1);
let (start_x, start_y, end_x, end_y) = match position {
  Position::HorizontalTop => (left, top, right, top),
  Position::HorizontalBottom => (left, bottom, right, bottom),
  Position::VerticalLeft => (left, top, left, bottom),
  Position::VerticalRight => right, top, right, bottom),
};

index = 0;
for x in (start_x..end_x) { // either this line
  for y in (start_y..end_y) { // or this line will just be a single iteration
    if index < start_track { draw_start_arrow(x, y); }
    else if index < start_thumb { draw_track(x, y); }
    else if index < end_thumb { draw_thumb(x, y); }
    else if index < end_track { draw_track(x, y); }
    else { draw_end_arrow(x, y); }
    index++;
  }
}
```

If I have indeed missed some complexity that is non obvious here, I apologize.

kdheepak commented:
I guess the cross is confusing naming. 

I made two changes based on your most recent comment:

1. I changed all uses for `size` to `start` and `end`. Previously I was using the `main_position` and `main_size` to render, now it uses `start` and `end` to render. Maybe that'll make it a little clearer?
2. I renamed all the variables to more accurately match the exact part of the scrollbar they are used for, i.e. `track`, `thumb`, etc. (What was previously `cross` is now `track_axis`, and `main_start` is now `track_start`. And there's a corresponding `track_end`. Similarly there's also a `thumb_start` and `thumb_end`.)

```
        //
        // For ScrollbarOrientation::VerticalRight
        //
        //                   ┌───────── track_axis  (x)
        //                   v
        //   ┌───────────────┐
        //   │               ║<──────── track_start (y1)
        //   │               █
        //   │               █
        //   │               ║
        //   │               ║<──────── track_end   (y2)
        //   └───────────────┘
        //
        // For ScrollbarOrientation::HorizontalBottom
        //
        //   ┌───────────────┐
        //   │               │
        //   │               │
        //   │               │
        //   └═══███═════════┘<──────── track_axis  (y)
        //    ^             ^
        //    │             └────────── track_end   (x2)
        //    │
        //    └──────────────────────── track_start (x1)
        //

```

mindoodoo commented on examples/scrollbar.rs:
Imo it would be interesting to have the keybindings for the scrolling be displayed at the top of the example.

kdheepak replied:
I added the keybindings at the top of the example:

<img width="748" alt="image" src="https://github.com/tui-rs-revival/ratatui/assets/1813121/0af7f3c5-7e07-4c5a-a3e8-d94c146506f7">

mindoodoo commented on src/widgets/scrollbar.rs:
```diff
@@ -0,0 +1,853 @@
+/// Scrollbar widget for tui-rs library.
+///
+/// This widget can be used to display a scrollbar in a terminal user interface.
+/// The following components of the scrollbar are customization in symbol and style.
```
```suggestion
/// The following components of the scrollbar are customizable in symbol and style.
```

mindoodoo commented on src/widgets/scrollbar.rs:
```diff
@@ -0,0 +1,853 @@
+    /// Only sets begin_symbol and end_symbol if they are already contain a value.
+    /// If begin_symbol and/or end_symbol were set to `None` explicitly, this function will respect
+    /// that choice.
+    pub fn symbol(mut self, symbol: Set) -> Self {
```
```suggestion
    pub fn symbols(mut self, symbol: Set) -> Self {
```

kdheepak replied:
Resolved in https://github.com/tui-rs-revival/ratatui/pull/228/commits/61da1cc3ee57951f786cd67af8ef34ddf2273513.

mindoodoo commented on src/widgets/scrollbar.rs:
```diff
@@ -0,0 +1,853 @@
+    /// │  └──────── thumb
+    /// └─────────── begin
+    ///
+    /// Only sets begin_symbol and end_symbol if they are already contain a value.
```
```suggestion
    /// Only sets begin_symbol and end_symbol if they already contain a value.
```

mindoodoo commented on src/widgets/scrollbar.rs:
```diff
@@ -0,0 +1,853 @@
+        }
+    }
+
+    fn get_thumb_start_end(
+        &self,
+        state_content_length: u16,
+        state_position: u16,
+        viewport_content_length: u16,
+        track_start: u16,
+        track_end: u16,
+    ) -> (u16, u16) {
+        let scroll_position_ratio = (state_position as f64 / state_content_length as f64).min(1.0);
+
+        let thumb_size = (((viewport_content_length as f64 / state_content_length as f64)
+            * (track_end - track_start) as f64)
+            .round() as u16)
+            .max(1);
```
```suggestion
        fn get_thumb_start_end(
        &self,
        state: &ScrollbarState,
        viewport_content_length: u16,
        track_start: u16,
        track_end: u16,
    ) -> (u16, u16) {
        let scroll_position_ratio = (state.position as f64 / state.content_length as f64).min(1.0);

        let thumb_size = (((viewport_content_length as f64 / state.content_length as f64)
            * (track_end - track_start) as f64)
            .round() as u16)
            .max(1);
```

Againt nit pick, but I'd simplify the arg list a tad bit

kdheepak replied:
Resolved in https://github.com/tui-rs-revival/ratatui/pull/228/commits/2284b202bd71c2e1c0549aaa37a89276aeee2c42.

mindoodoo commented on src/widgets/scrollbar.rs:
```diff
@@ -0,0 +1,853 @@
+            return;
+        }
+
+        let (thumb_start, thumb_end) = self.get_thumb_start_end(
+            state.content_length,
+            state.position,
```
```suggestion
        let (thumb_start, thumb_end) = self.get_thumb_start_end(
            &state
```

(if the previous suggestion is applied)

kdheepak replied:
Resolved in https://github.com/tui-rs-revival/ratatui/pull/228/commits/2284b202bd71c2e1c0549aaa37a89276aeee2c42.

mindoodoo reviewed (approved):
Awesome stuff guys, looks great !!

Only a couple nit picks, not blocking per say. Looking forward to seeing this in the next release ! :rocket:

joshka reviewed (approved):
LGTM

joshka commented:
To make this ready to merge, I'm going to push a rebase squash and cleanup the commit message
